src/orange/OrangeTrackView.hh

```diff
@@ -491,19 +556,81 @@ CELER_FUNCTION void OrangeTrackView::set_dir(Real3 const& newdir)
         {
             // The boundary crossing direction has changed! Reverse our plans
             // to change the logical state and move to a new volume.
-            states_.boundary[thread_]
-                = flip_boundary(states_.boundary[thread_]);
+            lsa.boundary() = flip_boundary(lsa.boundary());
         }
     }
 
     // Complete direction setting
-    states_.dir[thread_] = newdir;
+    lsa.dir() = newdir;
 
     this->clear_next_step();
 }
 
 //---------------------------------------------------------------------------//
 // PRIVATE MEMBER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Iterate over levels 1 to N to find the next step.
+ *
+ * Caller is responsible for finding the canidate next step on level 0, and
+ * passing the resultant Intersection object as an argument
+ */
+CELER_FUNCTION void
+OrangeTrackView::find_next_step_impl(detail::Intersection isect)
+{
+    // Zero for top-level universe
+    UniverseId min_uid{0};
+
+    // Find the nearest intersection from level 0 to current level inclusive,
+    // prefering the higher level (i.e., lowest uid)
+    for (auto levelid : range(LevelId{1}, states_.level[thread_] + 1))
+    {
+        auto lsa = this->make_lsa(levelid);
+        auto tracker = this->make_tracker(lsa.universe());
+        auto local_isect = tracker.intersect(this->make_local_state(levelid),
+                                             isect.distance);
+        if (local_isect.distance < isect.distance)
+        {
+            isect = local_isect;
+            min_uid = lsa.universe();
+        }
+    }
+
+    next_step_ = isect.distance;
+
+    // If there is a valid next surface, convert it from local to global
+    if (isect)
+    {
+        detail::UnitIndexer ui(params_.unit_indexer_data);
+        next_surface_ = celeritas::detail::OnSurface(
```

Comment:
Just as a note, for the surface crossing implementation we'll probably need to store the local surface index as well.